### PR TITLE
storage: introduce the notion of snapshot strategies

### PR DESCRIPTION
**Pull request by nvb**
In #16954 we'll be introducing a new approach to sending Range snapshots.
In order to accommodate that, we needed to introduce some mechanism for
senders and receivers of snapshots to distinguish between different kinds
of snapshots. This change lays the groundwork for this by introducing the
idea of a "snapshot strategy". At the moment, the only strategy is `KV_BATCH`,
but #16954 will be introducing a second one - `SST`.

This isn't intended to be merged until #16954 is ready as well. Still,
it's independent refactoring that can stand on its own, so I figured it
should be split into its own change.

Release note: None

### CODE REVIEW

File: pkg/storage/store_snapshot.go [hunk 1]
```diff
@@ -53,13 +51,250 @@ const (
...
+}
+
+// Send implements the snapshotStrategy interface.
+func (kvSS *kvBatchSnapshotStrategy) Send(
+	ctx context.Context,
+	stream outgoingSnapshotStream,
+	header SnapshotRequest_Header,
+	snap *OutgoingSnapshot,
+) error {
+	assertStrategy(header, SnapshotRequest_KV_BATCH)
```
Comment (tschottdorf) [inline pkg/storage/store_snapshot.go]: why not pass a context in here and use `log.Fatal`?

**Review by tbg — APPROVED** [review]
LGTM

I actually think it's fine to merge this right now. I find that introducing this abstraction with just one implementation still makes the code overall more legible, and I'm sure this will find some way to rot if you don't merge it for another few weeks. Perhaps sprinkle a comment that indicates that there will be a second impl, along with a link to the issue, so that if you never end up implementing the SST snapshots (unlikely) the casual reader has an idea of what's going on.

I didn't review the parts that looked like code movement in great detail.

**Comment by bdarnell** [comment]
:lgtm:

I'd rather go ahead and get the code movement merged instead of keeping this open while waiting for #16954. 

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24760#-:-LADdBvmDsm9ushYz5Rc:bfgmozf)*

File: pkg/storage/store_snapshot.go (hunk 1 above)
Comment (nvanbenschoten) [reply pkg/storage/store_snapshot.go]: Done.

**Pull request merged**